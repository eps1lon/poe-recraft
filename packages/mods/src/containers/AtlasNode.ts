import Mod from '../mods/Mod';
import { AtlasNodeProps, Tag } from '../schema';

import ImmutableContainer from './ImmutableContainer';

export const SEXTANT_RANGE = 55; // http://poecraft.com/atlas has 55

export interface Builder {
  mods: Mod[];
  props: AtlasNodeProps;
}
export type HumanId = string;

export default class AtlasNode extends ImmutableContainer<Mod, Builder> {
  public static build(props: AtlasNodeProps) {
    return new AtlasNode([], props);
  }

  public static withBuilder(builder: Builder): AtlasNode {
    return new AtlasNode(builder.mods, builder.props);
  }

  public props: AtlasNodeProps;

  constructor(mods: Mod[], props: AtlasNodeProps) {
    super(mods);

    this.props = props;
  }

  public builder(): Builder {
    return {
      mods: this.mods,
      props: this.props,
    };
  }

  /**
   * returns a list of nodes with sextant distance <= depth
   * sextant distance:
   * (a, b) => 1, if a.isInSextantRange(b)
   *           +inf, otherwise
   */
  public inSextantRange(atlas: AtlasNode[], max_depth = 1): AtlasNode[] {
    let expand: AtlasNode[] = [this];
    const in_range: Set<AtlasNode> = new Set();

    let depth = max_depth;

    // with every step mark every node in range as to be expanded and
    // decrement depth
    /* eslint-disable no-loop-func */
    while (expand.length && depth >= 0) {
      expand = expand.reduce((nodes, node) => {
        in_range.add(node);

        const new_nodes = atlas.filter(
          new_node =>
            node.isInSextantRange(new_node) &&
            node !== new_node &&
            !in_range.has(new_node),
        );

        return nodes.concat(new_nodes);
      }, [] as AtlasNode[]);

      depth -= 1;
    }
    /* eslint-enable no-loop-func */

    return Array.from(in_range.values());
  }

  public isInSextantRange(other: AtlasNode): boolean {
    return this.distance(other) <= SEXTANT_RANGE;
  }

  public distance(other: AtlasNode): number {
    return Math.sqrt(
      (this.props.x - other.props.x) ** 2 + (this.props.y - other.props.y) ** 2,
    );
  }

  public pos(): string {
    return `X: ${this.props.x} Y: ${this.props.y}`;
  }

  public getAllMods(atlas: AtlasNode[]): Mod[] {
    return this.inSextantRange(atlas).reduce(
      (mods, other) => mods.concat(other.mods),
      this.mods,
    );
  }

  public getTags(): Tag[] {
    return super
      .getTags()
      .concat(this.props.world_area.tags)
      .filter(
        // unique by id
        (tag, i, tags) => tags.findIndex(other => other === tag) === i,
      );
  }

  public maxModsOfType(): number {
    return Number.POSITIVE_INFINITY;
  }

  public inDomainOf(mod_domain: number): boolean {
    return mod_domain === Mod.DOMAIN.ATLAS;
  }

  public level(): number {
    return this.props.world_area.area_level;
  }

  public affectingMods(atlas: AtlasNode[]): Mod[] {
    return atlas.reduce((mods, other) => {
      if (this.isInSextantRange(other)) {
        return mods.concat(other.mods);
      } else {
        return mods;
      }
    }, [] as Mod[]); // this.mods will be passed on atlas.reduce
  }

  public activeMods(atlas: AtlasNode[]): Mod[] {
    return this.affectingMods(atlas).filter(
      mod => mod.spawnweightFor(this) > 0,
    );
  }

  public inactiveMods(atlas: AtlasNode[]): Mod[] {
    return this.affectingMods(atlas).filter(
      mod => mod.spawnweightFor(this) <= 0,
    );
  }
}
