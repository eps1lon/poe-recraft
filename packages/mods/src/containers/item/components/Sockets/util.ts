import Item from '../../Item';
import { Socket, SocketColor } from './types';
import ValueRange from '../../../../calculator/ValueRange';

/**
 * creates a Socket according to the highest attr requirement
 * if two or more requirements are equal the color is chosen
 * with the priority Green > Red > Blue
 * @param item
 */
export function rollRequirementBiasedStable(item: Item): Socket {
  const { dex, int, str } = item.requirements.list();
  const color_by_attr = [
    { value: dex.value, color: SocketColor.green },
    { value: str.value, color: SocketColor.red },
    { value: int.value, color: SocketColor.blue },
    // descending
  ].sort((a, b) => -ValueRange.cmpMax(a.value, b.value));

  return {
    color: color_by_attr[0].color,
  };
}
