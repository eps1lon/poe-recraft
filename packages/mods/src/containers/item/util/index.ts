export { default as AugmentableValue } from './AugmentableValue';
