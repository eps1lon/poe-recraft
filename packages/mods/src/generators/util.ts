import { GeneratorDetails } from './Generator';

// filter nothing
const allowAll = () => true;

export interface WithSpawnchance {
  spawnchance: number;
}

export const calculateSpawnchance = (
  collection: Array<GeneratorDetails<any>>,
  filter: (details: GeneratorDetails<any>) => boolean = allowAll,
): Array<GeneratorDetails<any> & WithSpawnchance> => {
  const sum_spawnweight = collection.reduce((sum, details) => {
    if (details.spawnweight != null && filter(details)) {
      return sum + details.spawnweight;
    } else {
      return sum;
    }
  }, 0);

  return collection.map(details => {
    let spawnchance = 0;

    if (details.spawnweight != null && filter(details)) {
      spawnchance = details.spawnweight / sum_spawnweight;
    }

    return {
      ...details,
      spawnchance,
    };
  });
};
