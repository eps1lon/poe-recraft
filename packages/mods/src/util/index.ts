export { Flags, anySet } from './Flags';
export { default as MetaData } from './MetaData';
