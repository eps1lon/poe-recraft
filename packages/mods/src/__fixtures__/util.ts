import * as fs from 'fs';
import * as path from 'path';

import {
  AtlasNodeProps,
  BaseItemTypeProps,
  CraftingBenchOptionsProps,
  EssenceProps,
  ModProps,
} from '../schema';

import AtlasNode from '../containers/AtlasNode';
import Item from '../containers/item/Item';
import { Essence, MasterBenchOption } from '../generators';
import Mod from '../mods/Mod';
import PropsTable, { TableProps } from '../helpers/PropsTable';
import { Buildable } from '../interfaces/Buildable';

const json_cache: Partial<Tables> = {};

export const createTable = <P extends TableProps, T, A1>(
  file: keyof Tables,
  constructor: Buildable<P, T, A1>,
  arg1: A1,
) => {
  if (json_cache[file] == null) {
    const body = fs.readFileSync(
      path.join(__dirname, `../../data/${file}.json`),
    );

    const json = JSON.parse(body.toString());

    // @ts-ignore
    json_cache[file] = new PropsTable(json, constructor, arg1);
  }

  return json_cache[file];
};

interface Tables {
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>;
  craftingbenchoptions: PropsTable<
    CraftingBenchOptionsProps,
    MasterBenchOption,
    undefined
  >;
  essences: PropsTable<EssenceProps, Essence, ModProps>;
  items: PropsTable<BaseItemTypeProps, Item, undefined>;
  mods: PropsTable<ModProps, Mod, undefined>;
}

export const createTables = () => {
  const tables = {
    atlas: createTable('atlas', AtlasNode, undefined),
    craftingbenchoptions: createTable(
      'craftingbenchoptions',
      MasterBenchOption,
      undefined,
    ),
    items: createTable('items', Item, undefined),
    mods: createTable('mods', Mod, undefined),
  } as Tables;

  // @ts-ignore: its not undefined though?!
  tables.essences = createTable('essences', Essence, tables.mods.all());

  return tables;
};

export const byWorldAreaId = (id: string) => (node: AtlasNodeProps) =>
  node.world_area.id === id;

export const fromWorldAreaId = (
  id: string,
  atlas: PropsTable<AtlasNodeProps, AtlasNode, undefined>,
): AtlasNode => {
  return atlas.from(node => node.world_area.id === id);
};

export const byEssenceName = (name: string) => (essence: EssenceProps) =>
  essence.base_item_type.name === name;
