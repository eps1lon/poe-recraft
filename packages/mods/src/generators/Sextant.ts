import { BaseError } from 'make-error';

import AtlasNode from '../containers/AtlasNode';
import Mod from '../mods/Mod';
import { ModProps } from '../schema';
import { Flags, anySet } from '../util/Flags';

import Orb, { SpawnableFlags as BaseSpawnableFlags } from './Orb';

export interface SpawnableFlags extends BaseSpawnableFlags {
  no_adjacents_with_spawnweight: boolean;
}
export type SpawnableFlag = keyof SpawnableFlags;

export interface ApplicableFlags extends Flags {
  wrong_tier_group: boolean;
}
export type ApplicableFlag = keyof ApplicableFlags;

export class ContextUndefined extends BaseError {
  constructor(context: string) {
    super(`context not set, set ${context}`);
  }
}

export class CorruptedState extends BaseError {
  constructor(message: string) {
    super(`corrupted state: ${message}`);
  }
}

export enum Type {
  apprentice = 1,
  journeyman = 2,
  master = 3,
}

export default class Sextant extends Orb<AtlasNode> {
  public static type = Type;

  public static modFilter(mod: ModProps): boolean {
    return super.modFilter(mod) && mod.domain === Mod.DOMAIN.ATLAS;
  }

  public static build(mods: ModProps[]): Sextant {
    return new Sextant(this.buildMods(mods));
  }

  /**
   * creates a list of mod that are either
   * a) already applied to this node because they are directly in range
   * b) already applied to adjacent nodes and would create duplicates if applied
   *    to this node
   * in other words: consider every mod from maps within a sextant distance of 2
   * @param {AtlasNode} node
   * @param {AtlasNode[]} atlas
   */
  public static blockedMods(target: AtlasNode, atlas: AtlasNode[]): Mod[] {
    return target
      .inSextantRange(atlas, 1)
      .reduce(
        (mods, secondary) => mods.concat(secondary.getAllMods(atlas)),
        [] as Mod[],
      );
  }

  public atlas: AtlasNode[] | undefined;
  public type: Type = Type.master;

  public applyTo(node: AtlasNode): AtlasNode {
    if (!anySet(this.applicableTo(node))) {
      const rolled = super.rollMod(node.removeAllMods());

      if (rolled !== node) {
        // something change
        // reset the context to signal that it has to be mutated
        this.atlas = undefined;
      }

      return rolled;
    } else {
      return node;
    }
  }

  // applicable to any node for now
  public applicableTo(node: AtlasNode): ApplicableFlags {
    const applicable_flags = {
      wrong_tier_group: false,
    };

    const area_ranges = {
      [Type.apprentice]: 72,
      [Type.journeyman]: 77,
      [Type.master]: 84,
    };

    applicable_flags.wrong_tier_group =
      node.props.world_area.area_level > area_ranges[this.type];

    return applicable_flags;
  }

  // I'm not sure how it determines the spawnweight
  // For example a map with no_monster_packs can still roll
  // a mod with spawnweight 0 for no_monster_packs if adjacent maps can roll it
  // I don't know which spawnweight it chooses though if
  // it can choose between multiple weights
  // for now we just search the adjacents if its 0 and pick first we find
  public findAdjacentWithSpawnweight(
    mod: Mod,
    node: AtlasNode,
  ): AtlasNode | undefined {
    if (this.atlas == null) {
      throw new ContextUndefined('atlas');
    }

    return node.inSextantRange(this.atlas).find(other => {
      return mod.spawnweightFor(other) > 0;
    });
  }

  public isModSpawnableOn(mod: Mod, node: AtlasNode): SpawnableFlags {
    const spawnable_flags = {
      ...super.isModSpawnableOn(mod, node),
      no_adjacents_with_spawnweight: false,
    };

    if (spawnable_flags.spawnweight_zero) {
      const adjacent_with_spawnweight = this.findAdjacentWithSpawnweight(
        mod,
        node,
      );

      if (adjacent_with_spawnweight === undefined) {
        spawnable_flags.no_adjacents_with_spawnweight = true;
      } else {
        spawnable_flags.spawnweight_zero = false;
      }
    }

    return spawnable_flags;
  }

  public modsFor(node: AtlasNode, whitelist: string[] = []) {
    if (this.atlas == null) {
      throw new ContextUndefined('atlas');
    }

    const blocked = Sextant.blockedMods(node, this.atlas);

    return super
      .modsFor(node, whitelist)
      .filter(({ mod }) => {
        const is_blocked = blocked.includes(mod);

        return !is_blocked;
      })
      .map(({ spawnable, spawnweight, mod, ...details }) => {
        if (spawnable === undefined || spawnweight === undefined) {
          throw new Error('spawnable is undefined');
        }

        let spawnweight_with_adjacents: number = spawnweight;

        if (spawnweight === 0 && !spawnable.spawnweight_zero) {
          // see Sextant.isModSpawnableOn
          // we took the spawnweight from an adjacent node
          const adjacent_with_spawnweight = this.findAdjacentWithSpawnweight(
            mod,
            node,
          );

          // this shouldn't be happening spawnweight should be > 0
          // if the flag is not set. only exception is whitelist
          const corrupted_state =
            adjacent_with_spawnweight == null &&
            !whitelist.includes('spawnweight_zero');

          if (corrupted_state) {
            throw new CorruptedState(
              'spawnweight should be > 0 if spawnweight_zero is false',
            );
          } else if (adjacent_with_spawnweight != null) {
            // we need to assure Flow that whitelist.includes didnt
            // mutate adjacent_with_spawnweight although we wouldnt care about
            // it because we dont care if it went from null > any
            // we just need to be sure that it didnt go from any > null
            // and this is given since it's the first check in the expression
            spawnweight_with_adjacents = mod.spawnweightFor(
              adjacent_with_spawnweight,
            );
          }
        }

        return {
          mod,
          spawnable,
          spawnweight: spawnweight_with_adjacents,
          ...details,
        };
      });
  }
}
