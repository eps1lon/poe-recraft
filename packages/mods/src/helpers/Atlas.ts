import AtlasNode from '../containers/AtlasNode';
import Sextant from '../generators/Sextant';
import Mod from '../mods/Mod';
import { AtlasNodeProps } from '../schema';
import { GeneratorDetails } from '../generators/Generator';

export type HumanId = string;
export type AtlasNodes = Map<HumanId, AtlasNode>;
export interface Builder {
  nodes: AtlasNodes;
}

type AtlasSextant = Sextant; // sextant.atlas defined

/**
 * immutable data structure for the atlas in Path of Exile
 *
 * main purpose is for reducer like usage in redux
 */
export default class Atlas {
  public static buildLookupTable(atlas: AtlasNodeProps[]): AtlasNodes {
    return atlas.reduce((nodes, props) => {
      return nodes.set(props.world_area.id, new AtlasNode([], props));
    }, new Map());
  }

  public static build(atlas: AtlasNodeProps[]): Atlas {
    const nodes: AtlasNodes = Atlas.buildLookupTable(atlas);

    return new Atlas(nodes);
  }

  public static withBuilder(builder: Builder): Atlas {
    return new Atlas(builder.nodes);
  }

  public readonly nodes: AtlasNodes;

  constructor(nodes: AtlasNodes) {
    this.nodes = nodes;
  }

  // TODO in the future use [Symbol.iterator]()
  public asArray(): AtlasNode[] {
    return Array.from(this.nodes.values());
  }

  /**
   * wrapper for map get that ensures a node or throws
   */
  public get(world_area_id: string): AtlasNode {
    const node = this.nodes.get(world_area_id);

    if (node == null) {
      throw new Error(`IndexError: '${world_area_id}' not found`);
    }

    return node;
  }

  public builder(): Builder {
    return {
      nodes: this.nodes,
    };
  }

  /**
   * batch mutations
   *
   * if the returned object is strict equal to the prev one
   * it doesn't return a new copy
   */
  public withMutations(mutate: (builder: Builder) => Builder): this {
    const prev = this.builder();
    const next = mutate(prev);

    if (prev !== next) {
      // @ts-ignore
      return this.constructor.withBuilder(next);
    } else {
      return this;
    }
  }

  /**
   * removes mods on all maps
   *
   * always returns a new copy
   */
  public reset(): this {
    return this.withMutations(({ nodes, ...builder }) => {
      return {
        ...builder,
        nodes: Atlas.buildLookupTable(
          this.asArray().map(node => node.removeAllMods().props),
        ),
      };
    });
  }

  public addMod(mod: Mod, world_area_id: string): this {
    return this.mutateNode(world_area_id, node => node.addMod(mod));
  }

  public removeMod(mod: Mod, world_area_id: HumanId): this {
    return this.mutateNode(world_area_id, node => node.removeMod(mod));
  }

  public mutateNode(
    world_area_id: string,
    mutate: (node: AtlasNode) => AtlasNode,
  ): this {
    const target = this.get(world_area_id);
    const mutated = mutate(target);

    if (target === mutated) {
      return this;
    } else {
      return this.withMutations(({ nodes, ...builder }) => {
        return {
          ...builder,
          nodes: new Map(nodes).set(world_area_id, mutated),
        };
      });
    }
  }

  public applySextant(sextant: Sextant, world_area_id: string): this {
    const sextant_on_atlas = this.prepareSextant(sextant);

    return this.mutateNode(world_area_id, node =>
      sextant_on_atlas.applyTo(node),
    );
  }

  public modsFor(
    sextant: Sextant,
    world_area_id: string,
  ): Array<GeneratorDetails<Mod>> {
    const sextant_on_atlas = this.prepareSextant(sextant);

    return sextant_on_atlas.modsFor(this.get(world_area_id));
  }

  public blockedMods(world_area_id: string): Mod[] {
    const target = this.get(world_area_id);

    return Sextant.blockedMods(target, this.asArray());
  }

  private prepareSextant(sextant: Sextant): AtlasSextant {
    const clone = new Sextant(sextant.mods);

    clone.type = sextant.type;
    clone.atlas = this.asArray();

    return clone;
  }
}
