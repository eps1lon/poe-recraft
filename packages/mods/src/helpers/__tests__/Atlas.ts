import { createTables } from '../../__fixtures__/util';
import AtlasNode from '../../containers/AtlasNode';
import Sextant from '../../generators/Sextant';

import Atlas from '../Atlas';

const { atlas: atlas_props, mods } = createTables();

it('should build', () => {
  const atlas = Atlas.build(atlas_props.all());

  expect(atlas).toBeInstanceOf(Atlas);
});

it('should have a human readable getter', () => {
  const atlas = Atlas.build(atlas_props.all());

  expect(atlas.get('MapWorldsDunes')).toBeInstanceOf(AtlasNode);
  expect(atlas.get('MapWorldsGardens')).toBeInstanceOf(AtlasNode);
  expect(() => atlas.get('UndefinedMap')).toThrow(
    "IndexError: 'UndefinedMap' not found",
  );
});

it('should return a new only if it changed', () => {
  const atlas = Atlas.build(atlas_props.all());

  expect(atlas.withMutations(b => b)).toBe(atlas);
  // created a new reference, not checking shallow equal
  expect(atlas.withMutations(b => ({ ...b }))).not.toBe(atlas);
});

it('should add mods', () => {
  const atlas = Atlas.build(atlas_props.all());
  const invasion_mod = mods.fromId('MapAtlasContainsAdditionalRandomBoss');

  const with_mods = atlas.addMod(invasion_mod, 'MapWorldsDunes');

  expect(with_mods).not.toBe(atlas);
  expect(atlas.get('MapWorldsDunes')).not.toBe(with_mods.get('MapWorldsDunes'));
  expect(with_mods.get('MapWorldsDunes').hasMod(invasion_mod)).toBe(true);

  // node didnt change => atlas didnt change
  expect(with_mods.addMod(invasion_mod, 'MapWorldsDunes')).toBe(with_mods);
});

it('should remove mods', () => {
  const invasion_mod = mods.fromId('MapAtlasContainsAdditionalRandomBoss');
  const atlas = Atlas.build(atlas_props.all()).addMod(
    invasion_mod,
    'MapWorldsDunes',
  );

  const without_mods = atlas.removeMod(invasion_mod, 'MapWorldsDunes');

  expect(without_mods).not.toBe(atlas);
  expect(without_mods.get('MapWorldsDunes')).not.toBe(
    atlas.get('MapWorldsDunes'),
  );
  expect(without_mods.get('MapWorldsDunes').hasMod(invasion_mod)).toBe(false);

  // node didnt change => atlas didnt change
  expect(without_mods.removeMod(invasion_mod, 'MapWorldsDunes')).toBe(
    without_mods,
  );
});

it('should be resettable', () => {
  const invasion_mod = mods.fromId('MapAtlasContainsAdditionalRandomBoss');
  const magick_packs_mod = mods.fromId('MapAtlasMagicPackSize');
  const atlas = Atlas.build(atlas_props.all())
    .addMod(invasion_mod, 'MapWorldsDunes')
    .addMod(magick_packs_mod, 'MapWorldsArcade');

  // pre
  expect(atlas.get('MapWorldsDunes').hasMod(invasion_mod)).toBe(true);
  expect(atlas.get('MapWorldsArcade').hasMod(magick_packs_mod)).toBe(true);

  const without_mods = atlas.reset();

  // post
  expect(without_mods).not.toBe(atlas);

  expect(without_mods.get('MapWorldsDunes').mods).toHaveLength(0);
  expect(without_mods.get('MapWorldsArcade').mods).toHaveLength(0);

  // reset always changes
  expect(without_mods.reset).not.toBe(without_mods);
});

it('should calc mods from sextants', () => {
  const atlas = Atlas.build(atlas_props.all());
  const sextant = Sextant.build(mods.all());

  const available_mods = atlas
    .modsFor(sextant, 'MapWorldsDunes')
    .map(({ mod }) => mod.props.id);

  expect(available_mods).toMatchSnapshot();
});

it('should apply sextants', () => {
  const atlas = Atlas.build(atlas_props.all());
  const sextant = Sextant.build(mods.all());

  const crafted = atlas.applySextant(sextant, 'MapWorldsDunes');

  // post
  expect(crafted.get('MapWorldsDunes').mods).toHaveLength(1);
  expect(crafted).not.toBe(atlas);

  sextant.type = Sextant.type.apprentice;
  // apprentice on red map
  expect(atlas.applySextant(sextant, 'MapWorldsPlaza')).toBe(atlas);
});
