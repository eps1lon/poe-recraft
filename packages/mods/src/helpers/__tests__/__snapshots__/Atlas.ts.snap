// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should calc mods from sextants 1`] = `
Array [
  "MapAtlasBossAdditionalUnique",
  "MapAtlasBossDropsCorruptedItems",
  "MapAtlasUnIDedMapsGrantQuantity",
  "MapAtlasContainsMaster",
  "MapAtlasContainsAdditionalRandomBoss",
  "MapAtlasMoveAttackCastSpeed",
  "MapAtlasFireDamageAndPacks",
  "MapAtlasColdDamageAndPacks",
  "MapAtlasLightningDamageAndPacks",
  "MapAtlasPhysicalDamageAndPacks_",
  "MapAtlasChaosDamageAndPacks_",
  "MapAtlasGroundHaste",
  "MapAtlasUniqueMonstersDropCorruptedItems",
  "MapAtlasMapsHave20PercentQuality",
  "MapAtlasMapQualityBonusAlsoAppliesToRarity",
  "MapAtlasPlayersDamageDealtAndTakenWhileStationary_",
  "MapAtlasOnslaughtWhenHitAndOnslaughtEffect",
  "MapAtlasContainsZanaMission",
  "MapAtlasContainsEinharMission__",
  "MapAtlasContainsAlvaMission",
  "MapAtlasContainsNikoMission",
  "MapAtlasStrongboxesAreCorruptedAndUnided",
  "MapAtlasMagicPackSize",
  "MapAtlasRogueExilesDamageAndDropAdditionalJewels",
  "MapAtlasMagicAndRareMapsHaveAdditionalPacks",
  "MapAtlasReducedReflectedDamageTakenMirroredRares",
  "MapAtlasClustersOfWealthyBarrels",
  "MapAtlasClustersOfVolatileBarrels",
  "MapAtlasClustersOfBloodwormBarrels",
  "MapAtlasClustersOfParasiteBarrels",
  "MapAtlasClustersOfBeaconBarrels",
  "MapAtlasStrongboxMonstersEnrageAndDrops",
  "MapAtlasPoisonEffectsAndParasites",
  "MapAtlasHighValueEnemiesOnOwnTeam",
  "MapAtlasMontersThatConvertOnDeath__",
  "MapAtlasInstantFlasksAndHealingMonsters",
  "MapAtlasBodyguardsAndBossMapDrops",
  "MapAtlasCurseEffectOnPlayersAndMonsters",
  "MapAtlasAdditionalBreaches",
  "MapAtlasAdditionalAbysses",
]
`;
